### PR TITLE
read_sas catches own error #24548

### LINKED ISSUE #24548: read_sas catches own error
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
pd.read_sas('/tmp/foo')
```

Output (`/tmp/foo` does not have to exist):

```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-10-de3f5c15fb71> in <module>
      1 import pandas as pd
----> 2 pd.read_sas('/tmp/foo')

~/.virtualenvs/pandas/lib/python3.7/site-packages/pandas/io/sas/sasreader.py in read_sas(filepath_or_buffer, format, index, encoding, chunksize, iterator)
     50             pass
     51 
---> 52     if format.lower() == 'xport':
     53         from pandas.io.sas.sas_xport import XportReader
     54         reader = XportReader(filepath_or_buffer, index=index,

AttributeError: 'NoneType' object has no attribute 'lower'
```

#### Problem description

When `format=None` the function attempts to infer the file's format by naively looking at the extension. If it is neither `'.xpt'` nor `'.sas7bdat'` then an error is raised. However, this logic is wrapped in a try/except block that catches the raised error and silently discards it:

https://github.com/pandas-dev/pandas/blob/179822a0a5b3e84c841bcb76e72a269d13cd3aad/pandas/io/sas/sasreader.py#L35-L53

The function then attempts to call `format.lower()` but `format` is still `None` so an `AttributeError` is raised.

#### Expected Output

Expected behaviour: error raised when unable to infer format. Suggest documenting that inference is done by checking the file's extension.

```python
ValueError: unable to infer format of SAS file
```

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.1.final.0
python-bits: 64
OS: Linux
OS-release: 4.19.12-arch1-1-ARCH
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: en_GB.UTF-8

pandas: 0.23.4
pytest: None
pip: 18.1
setuptools: 40.6.3
Cython: None
numpy: 1.15.4
scipy: None
pyarrow: None
xarray: None
IPython: 7.2.0
sphinx: None
patsy: None
dateutil: 2.7.5
pytz: 2018.7
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by benjaminr**
- [ ] closes #24548 
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
needs a test & a whatsnew entry

File: pandas/io/sas/sasreader.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ def read_sas(filepath_or_buffer, format=None, index=None, encoding=None,
             else:
                 raise ValueError("unable to infer format of SAS file")
         except ValueError:
-            pass
+            raise
```
Comment (samgd) [inline pandas/io/sas/sasreader.py]: https://github.com/pandas-dev/pandas/issues/24548#issuecomment-450906323

File: pandas/io/sas/sasreader.py [hunk 2]
```diff
@@ -48,7 +48,7 @@ def read_sas(filepath_or_buffer, format=None, index=None, encoding=None,
             else:
                 raise ValueError("unable to infer format of SAS file")
         except ValueError:
```
Comment (samgd) [inline pandas/io/sas/sasreader.py]: https://github.com/pandas-dev/pandas/issues/24548#issuecomment-450905723

**Review by samgd — CHANGES_REQUESTED** [review]

File: pandas/io/sas/sasreader.py (hunk 1 above)
Comment (jreback) [reply pandas/io/sas/sasreader.py]: yeah you can see if this is removable

Comment (benjaminr) [reply pandas/io/sas/sasreader.py]: Sure

File: pandas/io/sas/sasreader.py (hunk 2 above)
Comment (benjaminr) [reply pandas/io/sas/sasreader.py]: See latest commit.

File: doc/source/whatsnew/v0.24.0.rst [hunk 3]
```diff
@@ -1623,6 +1623,7 @@ Notice how we now instead output ``np.nan`` itself instead of a stringified form
 - Bug in :meth:`DataFrame.to_stata`, :class:`pandas.io.stata.StataWriter` and :class:`pandas.io.stata.StataWriter117` where a exception would leave a partially written and invalid dta file (:issue:`23573`)
 - Bug in :meth:`DataFrame.to_stata` and :class:`pandas.io.stata.StataWriter117` that produced invalid files when using strLs with non-ASCII characters (:issue:`23573`)
 - Bug in :class:`HDFStore` that caused it to raise ``ValueError`` when reading a Dataframe in Python 3 from fixed format written in Python 2 (:issue:`24510`)
+- Bug in :func:`read_sas()` in which ``ValueError`` supposedly raised by an inability to infer file format is passed instead of actually raised (:issue:`24548`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.rst]: this is not so clear, can you re-word

**Review by jreback — CHANGES_REQUESTED** [review]
minor comment, lgtm. ping on green.

File: doc/source/whatsnew/v0.24.0.rst (hunk 3 above)
Comment (benjaminr) [reply doc/source/whatsnew/v0.24.0.rst]: Sure, see latest commit.

**Comment by jreback** [comment]
i pushed a revision

**Comment by benjaminr** [comment]
@jreback Build appears to have mostly passed apart from Windows which is reporting:

"Could not find conda environment: pandas-dev" - not sure this is related to what I've changed.

**Comment by TomAugspurger** [comment]
restarted the azure build.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**